src/MessagePack.GeneratorCore/Generator/StringKey/EmbedStringHelper.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) All contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Text;
+
+namespace MessagePackCompiler.Generator
+{
+    public static class EmbedStringHelper
+    {
+        public static readonly Encoding Utf8 = new UTF8Encoding(false);
+
+        public static string ToByteArrayString(byte[] binary)
+        {
+            var headerLength = GetHeaderLength(binary.Length);
+            Span<byte> header = stackalloc byte[headerLength];
+            EmbedHeader(binary.Length, header);
+            var buffer = new StringBuilder().Append("new byte[").Append(headerLength).Append(" + ").Append(binary.Length).Append("] { ").Append(header[0]);
```

Comment:
This leads to generated code changing to a less efficient form:

```diff
-writer.WriteRaw(this.____stringByteKeys[1]);
+writer.WriteRaw(new byte[1 + 1] { 161, 97 });
```

This means instead of reusing a single array every time serialization happens, we allocate arrays every time. This seems like a step backwards.

Reply:
Thank you for your review!
Oh, I've missed the cases of the short length array.
[I examined those cases in sharplab.](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHAZgAJ9KBhSgb1MpeqpwBZKBZACgEpGzViIBKcCABMA8gDsMATwDKABwiyAPACEF8AHyUAHpQC8lWXADulYLrgBtALpCSIt6wIA2AkkoBOAHZfYXcAXwBuELcAdQQ0eFEIS15DfkjXd0pxKTlFVXVtOwMFU3MrGzsnF0y3H0o8X3IkKJEIltZY+LhE5IU09pZsmXllNU0dfUoAL1KLa1t4KqYMmpY6hsomyg5glbc2vZFOhKTeKf690KioimouY+7TodzRgom4A1So5ZqAMwB7BASADGAAtKLwAG4QBCUEpoWRGfgDaqrFg4Ah+Xh9dI1K6XUihIA==)

MessagePackWriter.WriteRaw has 2 overloads(ReadOnlySpan<byte> and ReadOnlySequence<byte>).
No array allocation seems to occur.

Reply:
I'm thinking about https://github.com/neuecc/MessagePack-CSharp/pull/1044#issuecomment-693029665 and the fact that mpc output is often used in Unity. Are we sure unity's compiler and il2cpp, etc. will provide adequate optimizations? As you left it in that PR, it sounds like you were not sure the optimizations will be there.

Reply:
I've examined Unity IL2CPP code for the static ReadOnlySpan&lt;byte&gt; optimization.

<details><summary>Unity C# source code</summary>

```csharp
using System;
using System.Globalization;
using TMPro;
using UnityEngine;

public class ArrayEmbed : MonoBehaviour
{
    [SerializeField] private TMP_Text text = default;
    
    void Start()
    {
        text.text = EmbedderHelper.Sum(EmbedderHelper.GetTemp()).ToString(CultureInfo.InvariantCulture);
    }
}


public static class EmbedderHelper
{
    public static ReadOnlySpan<byte> GetTemp() => new byte[1 + 1] { 161 , 12 };

    public static uint Sum(ReadOnlySpan<byte> span)
    {
        if (span.IsEmpty)
        {
            throw new ArgumentOutOfRangeException();
        }

        uint answer = span[0];
        foreach (var b in span.Slice(1))
        {
            answer += b;
        }

        return answer;
    }
}
```
</details>

<details><summary>The most important codes I extract.</summary>

```cpp
// System.ReadOnlySpan`1<System.Byte> EmbedderHelper::GetTemp()
extern "C" IL2CPP_METHOD_ATTR ReadOnlySpan_1_t11883B9D8CFA411A5B2BC1C72CCEB8215DFBE5E6  EmbedderHelper_GetTemp_m0A9B4121AE7DA48DDF6FA423281642A01BA2F505 (const RuntimeMethod* method)
{
	static bool s_Il2CppMethodInitialized;
	if (!s_Il2CppMethodInitialized)
	{
		il2cpp_codegen_initialize_method (EmbedderHelper_GetTemp_m0A9B4121AE7DA48DDF6FA423281642A01BA2F505_MetadataUsageId);
		s_Il2CppMethodInitialized = true;
	}
	{
		// public static ReadOnlySpan<byte> GetTemp() => new byte[1 + 1] { 161 , 12 };
		ReadOnlySpan_1_t11883B9D8CFA411A5B2BC1C72CCEB8215DFBE5E6  L_0;
		memset(&L_0, 0, sizeof(L_0));
		ReadOnlySpan_1__ctor_m7D8EDA9D5C232F7B9E48494FAEC1CD57865932E2((&L_0), (void*)(void*)(il2cpp_codegen_get_field_data(U3CPrivateImplementationDetailsU3E_t4846F3E0DC61BBD6AE249FF1EAAC4CB64097DE4A____42BD4AA0A1EA24FF3EB09F993F11256F6F424768_0_FieldInfo_var)), 2, /*hidden argument*/ReadOnlySpan_1__ctor_m7D8EDA9D5C232F7B9E48494FAEC1CD57865932E2_RuntimeMethod_var);
		return L_0;
	}
}
```

```cpp
inline const char* il2cpp_codegen_get_field_data(RuntimeField* field)
{
    return il2cpp::vm::Field::GetData(field);
}
```

```cpp
const char* Field::GetData(FieldInfo *field)
{
    if (field->type->attrs & FIELD_ATTRIBUTE_HAS_DEFAULT)
    {
        const Il2CppType* type = NULL;
        return Class::GetFieldDefaultValue(field, &type);
    }
    else if (field->type->attrs & FIELD_ATTRIBUTE_HAS_FIELD_RVA)
    {
        IL2CPP_NOT_IMPLEMENTED_NO_ASSERT(Field::GetData, "This works for array initialization data. Revisit any other RVA use case.");
        const Il2CppType* type = NULL;
        return Class::GetFieldDefaultValue(field, &type);
    }
    else
    {
        return NULL;
    }
}
```

```cpp
const char* Class::GetFieldDefaultValue(const FieldInfo *field, const Il2CppType** type)
{
    const Il2CppFieldDefaultValue *entry = MetadataCache::GetFieldDefaultValueForField(field);
    if (entry != NULL)
    {
        *type = MetadataCache::GetIl2CppTypeFromIndex(entry->typeIndex);
        if (entry->dataIndex == kDefaultValueIndexNull)
            return NULL;

        return (const char*)MetadataCache::GetFieldDefaultValueDataFromIndex(entry->dataIndex);
    }
    return NULL;
}
```
</details>